plugins/kotlin/fir/src/org/jetbrains/kotlin/idea/completion/contributors/FirSuperMemberCompletionContributor.kt

```diff
@@ -3,38 +3,115 @@
 import org.jetbrains.kotlin.idea.completion.checkers.CompletionVisibilityChecker
 import org.jetbrains.kotlin.idea.completion.context.FirBasicCompletionContext
-import org.jetbrains.kotlin.idea.completion.context.FirNameReferencePositionContext
+import org.jetbrains.kotlin.idea.completion.context.FirSuperReceiverNameReferencePositionContext
 import org.jetbrains.kotlin.idea.completion.contributors.helpers.collectNonExtensions
 import org.jetbrains.kotlin.idea.completion.lookups.CallableInsertionOptions
 import org.jetbrains.kotlin.idea.completion.lookups.CallableInsertionStrategy
 import org.jetbrains.kotlin.idea.completion.lookups.detectImportStrategy
+import org.jetbrains.kotlin.name.Name
 import org.jetbrains.kotlin.psi.KtNamedFunction
 import org.jetbrains.kotlin.psi.KtSuperExpression
+import org.jetbrains.kotlin.utils.addIfNotNull
 
 internal class FirSuperMemberCompletionContributor(
     basicContext: FirBasicCompletionContext,
     priority: Int
-) : FirCompletionContributorBase<FirNameReferencePositionContext>(basicContext, priority) {
-    override fun KtAnalysisSession.complete(positionContext: FirNameReferencePositionContext) = with(positionContext) {
-        val superReceiver = explicitReceiver as? KtSuperExpression ?: return
+) : FirCompletionContributorBase<FirSuperReceiverNameReferencePositionContext>(basicContext, priority) {
+    override fun KtAnalysisSession.complete(positionContext: FirSuperReceiverNameReferencePositionContext) = with(positionContext) {
+        val superReceiver = positionContext.superExpression
         val expectedType = nameExpression.getExpectedType()
         val visibilityChecker = CompletionVisibilityChecker.create(basicContext, positionContext)
-        val possibleReceiverScope = superReceiver.getKtType()?.getTypeScope() ?: return
-        val nonExtensionMembers = collectNonExtensions(possibleReceiverScope, visibilityChecker, scopeNameFilter)
-        collectDelegateCallToSuperMember(superReceiver, nonExtensionMembers, expectedType)
+        val superType = superReceiver.getKtType() ?: return
+
+        val (nonExtensionMembers: Iterable<KtCallableSymbol>, namesNeedDisambiguation: Set<Name>) = if (superType !is KtIntersectionType) {
+            val scope = superType.getTypeScope() ?: return
+            collectNonExtensions(scope, visibilityChecker, scopeNameFilter).asIterable() to emptySet()
+        } else {
+            getSymbolsAndNamesNeedDisambiguation(superType.conjuncts, visibilityChecker)
+        }
+        collectCallToSuperMember(superReceiver, nonExtensionMembers, expectedType, namesNeedDisambiguation)
+        collectDelegateCallToSuperMember(superReceiver, nonExtensionMembers, expectedType, namesNeedDisambiguation)
+
+    }
+
+    private fun KtAnalysisSession.getSymbolsAndNamesNeedDisambiguation(
+        superTypes: List<KtType>,
+        visibilityChecker: CompletionVisibilityChecker
+    ): Pair<Iterable<KtCallableSymbol>, Set<Name>> {
```

Comment:
OK, I have changed the implementation to store both the super type and the callable symbol. Now it does not matter what I use here. After my change I am using a list since there is no need to check for duplicates (since member traversal should not return duplicates.)